fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/AvroDimensionRowParser.java

```diff
@@ -93,34 +95,45 @@ private boolean doesSchemaContainAllDimensionFields(Dimension dimension, Schema
      *
      * @throws IllegalArgumentException thrown if JSON object `fields` is not present
      */
-    public Set<DimensionRow> parseAvroFileDimensionRows(Dimension dimension, String avroFilePath)
+    public Stream<DimensionRow> parseAvroFileDimensionRows(Dimension dimension, String avroFilePath)
         throws IllegalArgumentException {
 
         // Creates an AVRO DatumReader object
         DatumReader<GenericRecord> datumReader = new GenericDatumReader<>();
 
         // Creates an AVRO DataFileReader object that reads the AVRO data file one record at a time
-        try (DataFileReader<GenericRecord> dataFileReader = new DataFileReader<>(new File(avroFilePath), datumReader)) {
+        try {
+            DataFileReader<GenericRecord> dataFileReader = new DataFileReader<>(new File(avroFilePath), datumReader);
             // Validate Schema
             if (!doesSchemaContainAllDimensionFields(dimension, dataFileReader.getSchema())) {
                 String msg = "The AVRO schema file does not contain all the configured dimension fields";
                 LOG.error(msg);
                 throw new IllegalArgumentException(msg);
             }
-
+            Function<GenericRecord, Map<String, String>> recordMapFunction = genericRecord -> dimension
+                    .getDimensionFields()
+                    .stream()
+                    .collect(
+                            Collectors.toMap(
+                                    DimensionField::getName,
+                                    dimensionField -> resolveRecordValue(
+                                            genericRecord, dimensionFieldNameMapper.convert(dimension, dimensionField))
+                            )
+                    );
+            Runnable fileCloser = () -> {
+                try {
+                    dataFileReader.close();
+                } catch (IOException e) {
+                    String msg = String.format("Error closing avro file, at the location %s", avroFilePath);
+                    LOG.error(msg, e);
+                    throw new IllegalArgumentException(msg, e);
+                }
+            };
             // Generates a set of dimension Rows after retrieving the appropriate fields
             return StreamSupport.stream(dataFileReader.spliterator(), false)
-                    .map(genericRecord -> dimension.getDimensionFields().stream()
-                            .collect(
-                                 Collectors.toMap(
-                                     DimensionField::getName,
-                                     dimensionField -> resolveRecordValue(
-                                             genericRecord, dimensionFieldNameMapper.convert(dimension, dimensionField))
-                                 )
-                            )
-                    )
+                    .map(recordMapFunction)
                     .map(dimension::parseDimensionRow)
```

Comment:
Before if we threw an `IllegalArgumentException` here it would be closed before the `catch` block.

Now if it happens (if I'm reading [the javadoc](http://docs.oracle.com/javase/8/docs/api/java/util/stream/BaseStream.html#onClose-java.lang.Runnable-) correctly) we don't end up closing the file. So I think we'll need a `try`/`catch` in here